### PR TITLE
API/DEPR: Remove +/- as setops for DatetimeIndex/PeriodIndex (GH9630)

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
xref #13777, deprecations put in place in #9630

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
- still have to further clean-up the tests a bit (due to the previous API the tests for set ops and numeric ops are a bit mixed)
- for subtraction of two `PeriodIndex` objects, this is currently raising a TypeError, but I think we could actually allow it? Although that maybe depends on what is decided in https://github.com/pydata/pandas/issues/13077

**Comment by jreback** [comment]
we need to wait for #13077 on ``Period-Period`` (so can discuss over there). Let's limit scope to the more useful DTI-DTI

File: pandas/tseries/tests/test_base.py [hunk 1]
```diff
@@ -378,17 +378,34 @@ def test_add_iadd(self):
...
+            expected1 = pd.date_range('1/1/2000', freq='D', periods=10, tz=tz)
+
+            rng2 = pd.date_range('1/1/2000', freq='D', periods=5, tz=tz)
+            other2 = pd.date_range('1/4/2000', freq='D', periods=5, tz=tz)
+            expected2 = pd.date_range('1/1/2000', freq='D', periods=8, tz=tz)
+
+            rng3 = pd.date_range('1/1/2000', freq='D', periods=5, tz=tz)
+            other3 = pd.DatetimeIndex([], tz=tz)
+            expected3 = pd.date_range('1/1/2000', freq='D', periods=5, tz=tz)
+
```
Comment (jreback) [inline pandas/tseries/tests/test_base.py]: test subtracting different lengths (ValueError) IIRC; this maybe already tested in another section though

Comment (jreback) [inline pandas/tseries/tests/test_base.py]: mean this for below (dti-dti)

File: pandas/tseries/tests/test_base.py [hunk 2]
```diff
@@ -466,6 +515,30 @@ def test_sub_isub(self):
             rng -= 1
             tm.assert_index_equal(rng, expected)
 
+    def test_sub_dti_dti(self):
```
Comment (jreback) [inline pandas/tseries/tests/test_base.py]: I *think* we already have dti-scalar (both w and w/o tz) covered?

File: pandas/tseries/index.py [hunk 3]
```diff
@@ -731,19 +731,39 @@ def _add_datelike(self, other):
...
         # subtract a datetime from myself, yielding a TimedeltaIndex
         from pandas import TimedeltaIndex
-        other = Timestamp(other)
-        if other is tslib.NaT:
-            result = self._nat_new(box=False)
-        # require tz compat
-        elif not self._has_same_tz(other):
-            raise TypeError("Timestamp subtraction must have the same "
-                            "timezones or no timezones")
+        if isinstance(other, DatetimeIndex):
```
Comment (sinhrks) [inline pandas/tseries/index.py]: can this condition move to ``base``?

File: pandas/tseries/tests/test_base.py (hunk 2 above)
Comment (jorisvandenbossche) [inline pandas/tseries/tests/test_base.py]: yes, that is handled in the test_sub_isub test above

File: pandas/tseries/tests/test_base.py (hunk 1 above)
Comment (jorisvandenbossche) [inline pandas/tseries/tests/test_base.py]: added a test for that (and added code to give this a nicer error message)

File: pandas/tseries/index.py (hunk 3 above)
Comment (jorisvandenbossche) [inline pandas/tseries/index.py]: I moved the condition to base, but left it here as well (to follow the same pattern as for the other ops: one method in base (eg `_add_delta`) which in the subclass handles both scalars as index)

File: doc/source/whatsnew/v0.19.0.txt [hunk 4]
```diff
@@ -932,14 +932,16 @@ New Behavior:
...
 previously performed set operations (set union and difference). This
 behaviour was already deprecated since 0.15.0 (in favor using the specific
 ``.union()`` and ``.difference()`` methods), and is now disabled. When
 possible, ``+`` and ``-`` are now used for element-wise operations, for
-example for concatenating strings  (:issue:`8227`, :issue:`14127`).
+example for concatenating strings or subtracting datetimes
+(:issue:`8227`, :issue:`14127`).
 
-Previous Behavior:
+Previous behavior:
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: we need to standardize here, I see @jorisvandenbossche and @sinhrks changing these!

IIRC I think we do more of ``Previous Behavior`` (e.g. capitalized), and american spelling (no ``Behaviour``). But should just pick one :)

File: pandas/tseries/index.py [hunk 5]
```diff
@@ -731,19 +731,43 @@ def _add_datelike(self, other):
...
+                                                  fill_value=tslib.iNaT)
         else:
-            i8 = self.asi8
-            result = i8 - other.value
-            result = self._maybe_mask_results(result, fill_value=tslib.iNaT)
+            raise TypeError("cannot subtract DatetimeIndex and {typ}"
+                            .format(typ=type(other).__name__))
         return TimedeltaIndex(result, name=self.name, copy=False)
 
+    def _sub_datelike_dti(self, other):
```
Comment (jreback) [inline pandas/tseries/index.py]: this is fine. but just want to check that our naming convention is for these types of routines is consistent.

File: doc/source/whatsnew/v0.19.0.txt (hunk 4 above)
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.0.txt]: yes, will do that in https://github.com/pydata/pandas/pull/14176

We indeed need to just make a choice between American or British :-)
(is capitalizing all words also an American habit? :-))

**Comment by jreback** [comment]
lgtm.

File: pandas/tseries/index.py (hunk 5 above)
Comment (jorisvandenbossche) [inline pandas/tseries/index.py]: for adding a Timedelta-like vs TimedeltaIndex, there are the method `_add_delta_td` and `_add_delta_tdi` which are called from `_add_delta`, so copied that naming convention

File: doc/source/whatsnew/v0.19.0.txt (hunk 4 above)
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: I think 2 capital words sets it off s bit more
using American spelling

Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.0.txt]: while proofreading the whatsnew, I was thinking of putting the 'New/Old Behavior' in bold, that also sets it off a bit more

Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: even better

**Pull request merged**